### PR TITLE
ilm: Update action count only on success

### PR DESCRIPTION
**Pull request by krisis**
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description
`minio_node_ilm_action_count_*` for all ILM actions were incorrectly incremented before those actions' results were known.

## How to test this PR?
1. Make bucket with versioning
2. Set up ILM rules like below, for e.g
```
{
  "Rules": [
    {
      "Expiration": {
        "Days": 1,
        "ExpiredObjectAllVersions": true
      },
      "ID": "cnao1tgd6vv1q12j82b0",
      "Status": "Enabled"
    }
  ]
}
```
3. Upload objects and wait for them to expire
4. Bring a few drives offline to simulate quorum unavailability resulting in expiring of objects to fail

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: cmd/data-scanner.go [hunk 1]
```diff
@@ -1245,8 +1236,25 @@ func applyExpiryOnNonTransitionedObjects(ctx context.Context, objLayer ObjectLay
...
 		opts.DeletePrefix = true
 	}
+	var (
+		dobj ObjectInfo
+		err  error
+	)
+	defer func() {
+		if err != nil {
+			return
+		}
```
Comment (harshavardhana) [inline cmd/data-scanner.go:1266]: Do we count failures as well?

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/data-scanner.go (hunk 1 above)
Comment (krisis) [reply cmd/data-scanner.go:1266]: These metrics are meant to count only successful ILM actions.

**Comment by krisis** [comment]
@harshavardhana don't merge this PR yet. I would like to fix the missing accounting for expiration of tiered objects too.

**Comment by shtripat** [comment]
With write quorum lost ILM rule removes all versions with rules set for `ExpiredObjectAllVersions` and metrics `minio_node_ilm_action_count_delete_all_versions_action` reflects the correct removed versions count value. Hope this is expected result and yes, MinIO shows the error `Error: Write quorum may be lost on pool: 0, set: 0, expected write quorum: 3 (*errors.errorString)` due to quorum loss.

File: cmd/data-scanner.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/data-scanner.go:1266]: I am saying why not count errors as well?

**Comment by krisis** [comment]
@shtripat thank you for testing this PR under quorum loss conditions. Let me see if it is related this PR.

File: cmd/data-scanner.go (hunk 1 above)
Comment (krisis) [reply cmd/data-scanner.go:1266]: Yes, we could. I shall work on that in a separate PR. This one addresses an existing accounting bug.

**Review by harshavardhana — APPROVED** [review]

**Review by vadmeste — APPROVED** [review]
LGTM

**Pull request merged**